phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java

```diff
@@ -1138,10 +1329,10 @@ private void doPre(ObserverContext<RegionCoprocessorEnvironment> c, BatchMutateC
           metricSource.updatePreIndexUpdateFailureTime(dataTableName,
               EnvironmentEdgeManager.currentTimeMillis() - start);
           metricSource.incrementPreIndexUpdateFailures(dataTableName);
-          // Remove all locks as they are already unlocked. There is no need to unlock them again later when
-          // postBatchMutateIndispensably() is called
-          removePendingRows(context);
-          context.rowLocks.clear();
+          // Re-acquire all locks since we released them before making index updates
+          // Removal of reference counts and locks for the rows of this batch will be
```

Comment:
All cleanup now happens in the `postBatchMutateIndispensably` function which is always called irrespective of whether the preBatchMutate() and doPre() which it internally calls succeed or throw an exception. That function works under the assumption that rows have been locked so that it can safely manipulate the batch mutate context object. Because we released the row locks before calling doPre() and acquire them again in the success path, we need to do the same in the exception path.

Reply:
Makes sense, thanks for the explanation.